### PR TITLE
feat:Add Polynomial Addition

### PR DESCRIPTION
**Pull request by Ankita19ms0010**
#### Checklist

- [ ] Added description of change
- [x] Added file name matches [File name guidelines](https://github.com/TheAlgorithms/C-Plus-Plus/blob/master/CONTRIBUTING.md#New-File-Name-guidelines)
- [x] Added tests and example, test must pass
- [ ] Relevant documentation/comments is changed or added
- [x] PR title follows semantic [commit guidelines](https://github.com/TheAlgorithms/C-Plus-Plus/blob/master/CONTRIBUTING.md#Commit-Guidelines)
- [ ] Search previous suggestions before making a new one, as yours may be a duplicate.
- [x] I acknowledge that all my contributions will be made under the project's license.

Notes: 

<a href="https://gitpod.io/#https://github.com/TheAlgorithms/C/pull/600"></a>

### CODE REVIEW

File: data_structures/poly_add.c [hunk 1]
```diff
@@ -0,0 +1,265 @@
...
+ * @details
+ * This code takes two polynomials as input
+ * and prints their sum using linked list.
+ * The polynomials must be in increasing or decreasing order of degree.
+ * Degree must be positive.
+ */
+#include<stdio.h>  // for io operations
+#include<stdlib.h>
+
+struct term {
```
Comment (kvedala) [inline data_structures/poly_add.c:15]: Structures should be documented as such:
```c
/**
 * @brief identifier for single-variable polynomial coefficients as a linked list
 */
struct term {
     double coef; /**< coefficient value */
     int pow;         /**< power of the polynomial term */
     struct term *next; /**< pointer to the successive term */
 };
```

File: data_structures/poly_add.c [hunk 2]
```diff
@@ -0,0 +1,268 @@
...
+	}
+}
+
+/**
+ * Test function 1
+ * Polynomial 1 is 5 x^2 + 3 x^1 + 2 x^0
+ * Polynomial 2 is 7 x^3 + 9 x^1 + 10 x^0
+ * Resultant polynomial is 7 x^3 + 5 x^2 + 12 x^1 + 12 x^0
+ */
+void test1(struct term *poly1, struct term *poly2, struct term *poly3)
```
Comment (Panquesito7) [inline data_structures/poly_add.c:170]:
```suggestion
static void test1(struct term *poly1, struct term *poly2, struct term *poly3)
```

File: data_structures/poly_add.c [hunk 3]
```diff
@@ -0,0 +1,268 @@
...
+    printf ("\n") ;
+}
+
+/**
+ * Test function 2
+ * Polynomial 1 is 3 x^5 + 1 x^4 + 2 x^3 + -2 x^1 + 5 x^0
+ * Polynomial 2 is 2 x^5 + 3 x^3 + 7 x^1 + 2 x^0
+ * Resultant polynomial is 5 x^5 + 1 x^4 + 5 x^3 + 5 x^1 + 7 x^0
+ */
+void test2(struct term *poly1, struct term *poly2, struct term *poly3)
```
Comment (Panquesito7) [inline data_structures/poly_add.c:197]:
```suggestion
static void test2(struct term *poly1, struct term *poly2, struct term *poly3)
```

File: data_structures/poly_add.c [hunk 4]
```diff
@@ -0,0 +1,268 @@
...
+    printf ("\n") ;
+}
+
+/**
+ * Test function 3
+ * Polynomial 1 is -12 x^0 + 8 x^1 + 4 x^3
+ * Polynomial 2 is 5 x^0 + -13 x^1 + 3 x^3
+ * Resultant polynomial is -7 x^0 + -5 x^1 + 7 x^3
+ */
+void test3(struct term *poly1, struct term *poly2, struct term *poly3)
```
Comment (Panquesito7) [inline data_structures/poly_add.c:229]:
```suggestion
static void test3(struct term *poly1, struct term *poly2, struct term *poly3)
```

File: data_structures/poly_add.c [hunk 5]
```diff
@@ -0,0 +1,268 @@
...
+	{
+		printf("%d x^%d", poly->coef, poly->pow);
+		poly = poly->next;
+		if(poly != NULL)
+			printf(" + ");
+	}
+}
+
+/**
+ * Test function 1
```
Comment (Panquesito7) [inline data_structures/poly_add.c:165]:
```suggestion
 * @brief Test function 1
 *
 * @details
```

File: data_structures/poly_add.c [hunk 6]
```diff
@@ -0,0 +1,268 @@
...
+ * Polynomial 1 is 5 x^2 + 3 x^1 + 2 x^0
+ * Polynomial 2 is 7 x^3 + 9 x^1 + 10 x^0
+ * Resultant polynomial is 7 x^3 + 5 x^2 + 12 x^1 + 12 x^0
```
Comment (Panquesito7) [inline data_structures/poly_add.c:168]:
```suggestion
 * Resultant polynomial is 7 x^3 + 5 x^2 + 12 x^1 + 12 x^0
 * @returns void
```

File: data_structures/poly_add.c [hunk 7]
```diff
@@ -0,0 +1,268 @@
...
+	display_polynomial(poly2);
+
+	poly_add(&poly3, poly1, poly2);  // Adding the two polynomials
+    printf ("\nResultant polynomial:\n") ;
+    display_polynomial(poly3);
+    printf ("\n") ;
+}
+
+/**
+ * Test function 2
```
Comment (Panquesito7) [inline data_structures/poly_add.c:192]:
```suggestion
 * @brief Test function 2
 *
 * @details
```

File: data_structures/poly_add.c [hunk 8]
```diff
@@ -0,0 +1,268 @@
...
+ * Polynomial 1 is 3 x^5 + 1 x^4 + 2 x^3 + -2 x^1 + 5 x^0
+ * Polynomial 2 is 2 x^5 + 3 x^3 + 7 x^1 + 2 x^0
+ * Resultant polynomial is 5 x^5 + 1 x^4 + 5 x^3 + 5 x^1 + 7 x^0
```
Comment (Panquesito7) [inline data_structures/poly_add.c:195]:
```suggestion
 * Resultant polynomial is 5 x^5 + 1 x^4 + 5 x^3 + 5 x^1 + 7 x^0
 * @returns void
```

File: data_structures/poly_add.c [hunk 9]
```diff
@@ -0,0 +1,268 @@
...
+	display_polynomial(poly2);
+
+	poly_add(&poly3, poly1, poly2);  // Adding the two polynomials
+    printf ("\nResultant polynomial:\n") ;
+    display_polynomial(poly3);
+    printf ("\n") ;
+}
+
+/**
+ * Test function 3
```
Comment (Panquesito7) [inline data_structures/poly_add.c:224]:
```suggestion
 * @brief Test function 3
 *
 * @details
```

File: data_structures/poly_add.c [hunk 10]
```diff
@@ -0,0 +1,268 @@
...
+ * Polynomial 1 is -12 x^0 + 8 x^1 + 4 x^3
+ * Polynomial 2 is 5 x^0 + -13 x^1 + 3 x^3
+ * Resultant polynomial is -7 x^0 + -5 x^1 + 7 x^3
```
Comment (Panquesito7) [inline data_structures/poly_add.c:227]:
```suggestion
 * Resultant polynomial is -7 x^0 + -5 x^1 + 7 x^3
 * @returns void
```

File: data_structures/poly_add.c [hunk 11]
```diff
@@ -0,0 +1,268 @@
...
+	display_polynomial(poly2);
+
+	poly_add(&poly3, poly1, poly2);  // Adding the two polynomials
+    printf ("\nResultant polynomial:\n") ;
+    display_polynomial(poly3);
+    printf ("\n") ;
+}
+
+/**
+ * Main function
```
Comment (Panquesito7) [inline data_structures/poly_add.c:253]:
```suggestion
 * @brief Main function
 * @returns 0 on exit
```

**Review by Panquesito7 — CHANGES_REQUESTED** [review]

File: data_structures/poly_add.c [hunk 12]
```diff
@@ -0,0 +1,278 @@
...
+ * @returns none
+ */
+void display_polynomial(struct term* poly)
+{
+	while(poly != NULL)
+	{
+		printf("%d x^%d", poly->coef, poly->pow);
+		poly = poly->next;
+		if(poly != NULL)
+			printf(" + ");
```
Comment (Panquesito7) [inline data_structures/poly_add.c:160]:
```suggestion
		if(poly != NULL) {
			printf(" + ");
		}
```

**Review by Panquesito7 — CHANGES_REQUESTED** [review]

**Review by Panquesito7 — APPROVED** [review]
LGTM. 👍 😄

**Comment by kvedala** [comment]
Please move the file to `misc` folder using the command:
```sh
$ git mv data_structures/poly_add.c misc/poly_add.c
$ git commit -m "file path fixed"
$ git push
```

**Review by kvedala — CHANGES_REQUESTED** [review]
All `temp` variable pointers with dynamically allocated memories must be freed. For example, line# 123 where new memory is incrementally allocated, is not freed before the function returns. Also, in `test` function, the dynamic memory allocated by the create_polynomial needs to be deallocated. 

Moreover, the deallocated by calling `free(poly1)` in main function is incorrect as it does not deallocate the memory of the chain. A simple solution would be to reverse the `create_polynomial` and deallocate memory in a chain:
```c
void free_poly(struct term* poly)
{
    if(!poly) /* NULL pointer does not need delete */
        return;
    free(poly->next);  /* ensure that next linked term is also deleted -> this will lead to recursive deletions as required. */
    free(poly); /* delete the current term */
}
```

**Comment by Ankita19ms0010** [comment]
@Panquesito7 , this pull request has been inactive for a few days now. Any additional changes?

**Review by Panquesito7 — APPROVED** [review]
Looks good to me, good work. 👍
It seems @kvedala has been inactive for more than a week.

**Pull request merged**